diff --git a/CMakeLists.txt b/CMakeLists.txt
index ceae8f39..f3636f4e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -332,6 +334,9 @@
 target_include_directories(uv_shared PRIVATE include src)
 target_link_libraries(uv_shared ${uv_libraries})
 set_target_properties(uv_shared PROPERTIES OUTPUT_NAME "uv")
+if(MSVC)
+  set_target_properties(uv_shared PROPERTIES IMPORT_SUFFIX "_import.lib")
+endif()
 
 add_library(uv_static STATIC ${uv_sources})
 target_compile_definitions(uv_static PRIVATE ${uv_defines})
@@ -360,7 +365,7 @@
            WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
 endif()
 
-if(UNIX)
+macro(install_libuv_pc)
   # Now for some gibbering horrors from beyond the stars...
   include(GNUInstallDirs)
   foreach(x ${uv_libraries})
@@ -368,10 +373,16 @@
   endforeach(x)
   file(STRINGS configure.ac configure_ac REGEX ^AC_INIT)
   string(REGEX MATCH [0-9]+[.][0-9]+[.][0-9]+ PACKAGE_VERSION "${configure_ac}")
+  configure_file(libuv.pc.in ${CMAKE_CURRENT_BINARY_DIR}/libuv.pc @ONLY)
+  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/libuv.pc
+          DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
+endmacro()
+
+if(UNIX)
   set(includedir ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR})
   set(libdir ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR})
   set(prefix ${CMAKE_INSTALL_PREFIX})
-  configure_file(libuv.pc.in ${CMAKE_CURRENT_BINARY_DIR}/libuv.pc @ONLY)
+  install_libuv_pc()
 
   install(DIRECTORY include/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
   install(FILES LICENSE DESTINATION ${CMAKE_INSTALL_DOCDIR})
@@ -379,4 +390,19 @@
           DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
   install(TARGETS uv_shared LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})
   install(TARGETS uv_static ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+elseif(WIN32)
+  if(MINGW)
+    install_libuv_pc()
+  endif()
+  install(DIRECTORY include/ DESTINATION ${CMAKE_INSTALL_PREFIX}/include)
+  install(FILES LICENSE DESTINATION ${CMAKE_INSTALL_PREFIX})
+  install(TARGETS uv_shared
+    RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/bin  # for dll/executable files
+    LIBRARY DESTINATION ${CMAKE_INSTALL_PREFIX}/lib  # for shared library
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_PREFIX}/lib) # for static library
+  install(TARGETS uv_static
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_PREFIX}/lib) # for static library
+  if(MSVC)
+    install(FILES $<TARGET_PDB_FILE:uv_shared> DESTINATION ${CMAKE_INSTALL_PREFIX}/bin OPTIONAL)
+  endif()
 endif()
